### ISSUE TITLE
Reg: Detection of word boundaries in a sentence

### ISSUE THREAD

**Issue by ellurunaresh** [cmusphinx/g2p-seq2seq#126]
Hi all,
I am having a raw text of sentences without word boundaries. I am trying to train model with the sentence to words using a g2p model instead of words to phonemes.
E.g. goodmorningtoall  - good morning to all
Can we train a g2p model using the examples of above dictionary? 
Please help me how to solve this issue?

**Comment by nurtas-m** [comment]
Hello, @ellurunaresh 

No, in my mind, the solution based on g2p-seq2seq don't fit to your problem. Because:
1) The vocabulary should be fixed and not too big. The left part in your samples will be splitted into letters, and the right part will be splitted into the words. So, the vocabulary file will consist of the English letters and all English words. It means, that if you apply g2p-seq2seq to your data, it will create Neural Network with output size equal to the size of the English words vocabulary. The more output neurons will be in your model the less accuracy you will get and the more data you need to train the neural network;

2) The size of the graphemes and phonemes sequence limited to 30 symbols. You can increase this parameter, but the accuracy of your model will drop.

In my opinion, you need to apply another project to solve this issue. For example, punctuator project will be more reliable to solve your task, in my mind: https://github.com/ottokart/punctuator2

**Comment by nurtas-m** [comment]
I recommend you to use punctuator, because your problem very similar to the tasks that punctuator project solves successfully. You simply need to assume the letters as words and punctuations as SPACE_CHR. So you may feed training dataset in following format:
g o o d SPACE_CHR m o r n i n g SPACE_CHR t o SPACE_CHR a l l
h e l l o SPACE_CHR t o SPACE_CHR e v e r y o n e
etc
And, the program will predict the space character after each letter. 

If you don't want to use punctuator, you may adopt your problem to one of the problems from tensor2tensor realization. g2p-seq2seq toolkit intended to solve only grapheme-to-phoneme conversions. And, also it fully based on tensor2tensor.

**Comment by nurtas-m** [comment]
Hello, @ellurunaresh 

> As you mentioned above, I have trained g2p seq2seq model instead punctuator project.
Sorry, I don't understand. You have trained punctuator project or g2p-seq2seq project?

**Comment by ellurunaresh** [comment]
Sorry I have trained g2p seq2seq model.

**Comment by nurtas-m** [comment]
So, as I understand, you have trained g2p-seq2seq model on the data preprocessed like:
`goodmorning g o o d SPACE m o r n i n g`
`hellotoeveryone h e l l o SPACE t o SPACE e v e r y o n e`
`etc.`

If you have got the model that correctly predict for word boundaries, you may just ignore the exact output characters of the decoder, and use only the information about SPACE character position.

> How to explain to the decoder RNN not to take unseen context ( d e p e m).

You don't need to explain anything to decoder RNN.

> Is it possible to use constrained vocabulary so that decoder RNN should not generate OOV pronunciations? ( OOV characters).

The decoder RNN in g2p-seq2seq toolkit can't generate OOV pronunciations. In the case you have trained the model on the previously described data format (`hitoeveryone h i SPACE t o SPACE e v e r y o n e`), your vocabulary will consist of only English characters and SPACE symbol. You can check generated "vocab.g2p" file in your model_dir. So, "p" character is presented in your vocabulary (that consist all English characters), it is not an OOV character.

> How difficult to modify the code of seq2seq.

It is very subjective.

So, you can just replace decodes in g2p.py module with its inputs sequence (not separated with spaces), and only add the space character (" ") at the positions where "SPACE" phonemes were found:

`inputs = ['a', 'p', 'r', 'i', 'l', 'd', 'e', '**c**', 'e', 'm', 'b', 'e', 'r']`
`decodes = ['a', 'p', 'r', 'i', 'l', 'SPACE', 'd', 'e', '**p**', 'e', 'm', 'b', 'e', 'r']`

Find all space positions:
`space_positions = [sym_pos for sym_pos, sym in enumerate(decodes) if sym == 'SPACE']`

Replace decodes with inputs and spaces:
`output_str = ""`
`for pos, sym in enumerate(inputs):`
`    if pos in space_positions:`
`        output_str += " "`
`    output_str += sym`
`print("Input:{}".format("".join(inputs)))`
`print("Output:{}".format(output_str))`

**Comment by ellurunaresh** [comment]
Thanks for the reply.
 I was doing on the test set and I don't have ground truth Information. Then, how can I stop the phoneme sequence of  "d e p e m b e r".
